crates/matrix-sdk/src/room/timeline/mod.rs

```diff
@@ -0,0 +1,217 @@
+        EventTimelineItem, Message, PaginationOutcome, ReactionDetails, TimelineDetails,
+        TimelineItemContent, TimelineKey,
+    },
+    virtual_item::VirtualTimelineItem,
+};
+
+/// A high-level view into a regular¹ room's contents.
+///
+/// ¹ This type is meant to be used in the context of rooms without a
+/// `room_type`, that is rooms that are primarily used to exchange text
+/// messages.
+#[derive(Debug)]
+pub struct Timeline {
+    inner: TimelineInner,
+    room: room::Common,
+    start_token: Mutex<Option<String>>,
+    _end_token: Mutex<Option<String>>,
+    _event_handler_guard: EventHandlerDropGuard,
+}
+
+#[derive(Clone, Debug, Default)]
+struct TimelineInner {
+    items: MutableVec<Arc<TimelineItem>>,
+    // Reaction event / txn ID => sender and reaction data
+    reaction_map: Arc<Mutex<HashMap<TimelineKey, (OwnedUserId, AnnotationRelation)>>>,
+}
+
+impl Timeline {
+    pub(super) fn new(room: &room::Common) -> Self {
+        let inner = TimelineInner::default();
+
+        let handle = room.add_event_handler({
+            let inner = inner.clone();
+            move |event, encryption_info: Option<EncryptionInfo>, room: Room| {
+                let inner = inner.clone();
+                async move {
+                    handle_live_event(event, encryption_info, room.own_user_id(), &inner);
+                }
+            }
+        });
+        let _event_handler_guard = room.client.event_handler_drop_guard(handle);
+
+        Timeline {
+            inner,
+            room: room.clone(),
+            start_token: Mutex::new(None),
+            _end_token: Mutex::new(None),
+            _event_handler_guard,
+        }
+    }
+
+    /// Add more events to the start of the timeline.
+    pub async fn paginate_backwards(&self, limit: UInt) -> Result<PaginationOutcome> {
+        let start = self.start_token.lock().unwrap().clone();
+        let messages = self
+            .room
+            .messages(assign!(MessagesOptions::backward(), {
+                from: start.as_deref(),
+                limit,
+            }))
+            .await?;
+
+        let outcome = PaginationOutcome { more_messages: messages.end.is_some() };
+        *self.start_token.lock().unwrap() = messages.end;
+
+        let own_user_id = self.room.own_user_id();
+        for room_ev in messages.chunk {
+            handle_back_paginated_event(
+                room_ev.event.cast(),
+                room_ev.encryption_info,
+                own_user_id,
+                &self.inner,
+            );
+        }
+
+        Ok(outcome)
+    }
+
+    /// Get a signal of the timeline's items.
+    ///
+    /// You can poll this signal to receive updates, the first of which will
+    /// be the full list of items currently available.
+    ///
+    /// See [`SignalVecExt`](futures_signals::signal_vec::SignalVecExt) for a
+    /// high-level API on top of [`SignalVec`].
+    pub fn signal(&self) -> impl SignalVec<Item = Arc<TimelineItem>> {
```

Comment:
Should we expose `SignalVec` here, if we just expose `impl Stream...` that would likely be more familiar to most people. Another thing, when I tried the API out myself, the stream was the nicer one to use I would say.

Do we lose much if we return the stream here instead?

Reply:
I don't think we lose much, but also *why not both*? I'll add a `.stream()` accessor.